theano/sandbox/mkl/mkl_type.py

```diff
@@ -0,0 +1,270 @@
+from __future__ import absolute_import, print_function, division
+import os
+import numpy
+import theano
+from theano import Type, Variable
+from theano import tensor
+from six import StringIO
+
+try:
+    import mkl_ndarray
+    import mkl_ndarray.mkl_ndarray as mkl
+
+    try:
+        mkl_ndarray.__file__
+    except AttributeError:
+        from theano.gof.cmodule import get_lib_extension
+        mkl_ndarray.__file__ = os.path.join(mkl_ndarray.__path__._path[0], 'mkl_ndarray.' + get_lib_extension())
+except ImportError:
+    mkl = None
+
+
+class _operators(tensor.basic._tensor_py_operators):
+    dtype = property(lambda s: s.type.dtype)
+    broadcastable = property(lambda s: s.type.broadcastable)
+    ndim = property(lambda s: s.type.ndim)
+
+
+class MKLNdarrayVariable(_operators, Variable):
+    pass
+
+
+class MKLNdarrayType(Type):
```

Comment:
I mean the first one, layout specifies relative location of elements in memory. 
Different MKL DNN operation (for example, convolution, pool, etc.,) might use different layout , and even if the same operation, it could have different layout in different layers or topologies according to their parameters (such as kernel/windows, stride, padding).

You're correct about F32/F64, there're two sets of functions for float32 and float64 currently, and more precision will be supported in future.